### PR TITLE
[SDKv2] Add Account support for secp256k1 scheme

### PR DESCRIPTION
**Pull request by 0xmigo**
### Description

Prev stack - https://github.com/aptos-labs/aptos-core/pull/10321

This PR updates the `Account` class to be compatible with `secp256k1` scheme.

1. Split the AuthenticationKey scheme into `SigningScheme` and `DeriveScheme`
2. Add Account support for `secp256k1` scheme
3. Update unit test

### Test Plan

pnpm fmt
pnpm test

### CODE REVIEW

File: ecosystem/typescript/sdk_v2/src/core/account.ts [hunk 1]
```diff
@@ -49,17 +56,41 @@ export class Account {
...
+    // Derive the signing scheme from the public key
+    if (this.publicKey instanceof Ed25519PublicKey) {
+      this.signingScheme = SigningScheme.Ed25519;
+    } else if (this.publicKey instanceof MultiEd25519PublicKey) {
+      this.signingScheme = SigningScheme.MultiEd25519;
+    } else if (this.publicKey instanceof Secp256k1PublicKey) {
+      // Secp256k1
+      this.signingScheme = SigningScheme.Secp256k1Ecdsa;
+    } else {
+      throw new Error("Unsupported public key type");
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/core/account.ts]:
```suggestion
      throw new Error("Can not create new Account, unsupported public key type");
```

File: ecosystem/typescript/sdk_v2/src/core/account.ts [hunk 2]
```diff
@@ -49,17 +56,41 @@ export class Account {
...
+  static generate(args: { scheme: SigningScheme }): Account {
+    let privateKey: PrivateKey;
+
+    if (args.scheme === SigningScheme.Ed25519) {
+      privateKey = Ed25519PrivateKey.generate();
+    } else if (args.scheme === SigningScheme.Secp256k1Ecdsa) {
+      privateKey = Secp256k1PrivateKey.generate();
+    } else {
+      // TODO: Add support for MultiEd25519
+      throw new Error("Unsupported signing scheme");
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/core/account.ts]:
```suggestion
      throw new Error(`Can not generate new Private Key, unsupported signing scheme ${scheme}`);
```

File: ecosystem/typescript/sdk_v2/src/core/account.ts [hunk 3]
```diff
@@ -49,17 +56,41 @@ export class Account {
...
    *
+   * @param args.scheme AuthenticationKeyScheme - type of Authentication Key schemes
    * @returns Account
    */
-  static generate(): Account {
-    const privateKey = Ed25519PrivateKey.generate();
+  static generate(args: { scheme: SigningScheme }): Account {
+    let privateKey: PrivateKey;
+
+    if (args.scheme === SigningScheme.Ed25519) {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/core/account.ts]: nit
`const {scheme} = args`

File: ecosystem/typescript/sdk_v2/src/crypto/authentication_key.ts [hunk 4]
```diff
@@ -75,9 +76,11 @@ export class AuthenticationKey {
 
     let scheme: number;
     if (publicKey instanceof Ed25519PublicKey) {
-      scheme = AuthenticationKeyScheme.Ed25519.valueOf();
+      scheme = SigningScheme.Ed25519.valueOf();
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/authentication_key.ts:79]: interesting, then the `valueOf()` is needed? I believe we use enum strings in different places without it.

File: ecosystem/typescript/sdk_v2/src/crypto/secp256k1.ts [hunk 5]
```diff
@@ -161,11 +161,11 @@ export class Secp256k1PrivateKey extends PrivateKey {
 
   /**
    * Derive the Secp256k1PublicKey from this private key.
-   * 
+   *
    * @returns Secp256k1PublicKey
    */
   publicKey(): Secp256k1PublicKey {
-    const bytes = secp256k1.getPublicKey(this.key.toUint8Array());
+    const bytes = secp256k1.getPublicKey(this.key.toUint8Array(), false);
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/secp256k1.ts:168]: why `false`?

File: ecosystem/typescript/sdk_v2/src/types/index.ts [hunk 6]
```diff
@@ -796,6 +800,16 @@ export enum AuthenticationKeyScheme {
    * For MultiEd25519PublicKey
    */
   MultiEd25519 = 1,
+  /**
+   * For Secp256k1 ecdsa
+   */
+  Secp256k1Ecdsa = 2,
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/types/index.ts:808]: where are the numbers coming from? is it set in the rust code? if so lets add a link to it, same as https://github.com/aptos-labs/aptos-core/blob/main/ecosystem/typescript/sdk_v2/src/types/index.ts#L58

File: ecosystem/typescript/sdk_v2/src/core/account.ts [hunk 7]
```diff
@@ -49,17 +56,41 @@ export class Account {
...
+    }
+
     this.privateKey = privateKey;
     this.accountAddress = address;
   }
 
   /**
    * Derives an account with random private key and address
    *
+   * @param args.scheme AuthenticationKeyScheme - type of Authentication Key schemes
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/core/account.ts]: lets also add explanation on what types Aptos currently supports

**Review by 0xmaayan — COMMENTED** [review]
I am interesting in the use of `valueOf()`, have you tested it is not working without it?

File: ecosystem/typescript/sdk_v2/src/crypto/secp256k1.ts (hunk 5 above)
Comment (0xmigo) [reply ecosystem/typescript/sdk_v2/src/crypto/secp256k1.ts:168]: Good question. If true, the secp256k1 lib will compress the public key into 32 bytes. Instead we want an un-compressed one, which is 65 bytes.

File: ecosystem/typescript/sdk_v2/src/types/index.ts (hunk 6 above)
Comment (0xmigo) [reply ecosystem/typescript/sdk_v2/src/types/index.ts:808]: Yep its defined on the rust side. Will do!

File: ecosystem/typescript/sdk_v2/src/crypto/authentication_key.ts (hunk 4 above)
Comment (0xmigo) [reply ecosystem/typescript/sdk_v2/src/crypto/authentication_key.ts:79]: I believe the `valueOf()` returns the actual value or `number` of that enum. But if string is needed, we can run `toString()` to get the string version.

**Comment by 0xmigo** [comment]
Merging this into base PR https://github.com/aptos-labs/aptos-core/pull/10321

**Pull request merged**